WIP: fix(atlasmap-ui): Show only active mapping specific error/warn

igarashitm opened the pull request:
Fixes: #112 

waiting for atlasmap/atlasmap-runtime#280 to be merged

deeleman commented on src/app/lib/atlasmap-data-mapper/services/mapping.management.service.ts:
```diff
@@ -316,25 +316,35 @@ export class MappingManagementService {
+                if (mapping) {
+                    mapping.validationErrors = activeMappingErrors;
+                }
                 console.log("Finished fetching and parsing validation errors in " + (Date.now() - startTime) + "ms.");
```
`console.log`might be useful for development purposes but exposes unwanted messages whilst in production and hinders further debugging of other issues. I've noticed that there're loggers like this scattered throughout the app. Perhaps this is a good moment to start removing them, just a one (or a handful) at a time ;)

deeleman commented on src/app/lib/atlasmap-data-mapper/services/mapping.serializer.ts:
```diff
@@ -41,6 +41,7 @@ export class MappingSerializer {
                     jsonMapping = {
                         "jsonType" : "io.atlasmap.v2.Mapping",
                         "mappingType" : "MAP",
+                        'id' : m.uuid,
```
Although it is a common practice in languages such as Java or C# when implementing iterables or lambda expressions, naming things with a single letter is discouraged in dynamically compiled languages such as JavaScript. The reason is no other than to ease type inference at a glance. Bear in mind that we currently use TypeScript, but in ES6 we do not have static type checking and figuring out types is hard.

How about repurposing the `for...of` loop above into:

```for (let mapping of mappingDefinition.mappings) {```

and then rename all the underlying references to the new name. This way any other developer will infer at a glance where are we getting that `uuid` from just reading the code. ;)

deeleman replied:
Also, consistency is key when naming things in loosely-typed languages. The parent `for...of` could be repurposed into:

```for (let fieldMapping of mapping.fieldMappings) { ...```

Again, the index name will inform us of the underlying type of that object. ;)

deeleman replied:
Also, such naming helps us understand that `mapping.fieldMappings` and `json.AtlasMapping.mappings.mapping` allocate the same typed collection so to say, so it's easier to undertand the different existing types at a glance. :)

igarashitm replied:
OK I think I cleaned up those not only around the fix, but entire files. One thing to note, the `mapping.fieldMappings` where `mapping` is an instance of `MappingModel`, is actually an instance of `FieldMappingPair[]`, which is yet another container held in the `MappingModel`. So I used `fieldMappingPair` as an instance name instead.

deeleman commented on src/app/lib/atlasmap-data-mapper/services/mapping.serializer.ts:
```diff
@@ -306,6 +307,9 @@ export class MappingSerializer {
+                m.uuid = fieldMapping.id;
+            }
             m.fieldMappings = [];
 
```
The use of `var` is discouraged in TypeScript applications. The convention is to replace its use completely by using either `let` whenever the variable is meant be reassigned or `const` to signal that such variable is immutable by reference. Also, naming something like `m` is not a good practice. Something like this would work better and ease future maintainability by other peer developers: 

```
for (let fieldMapping of json.AtlasMapping.mappings.mapping) {
   // Please note below that it is not necessary to annotate types in assignation expressions
  const mappingModel = new MappingModel();

  if (fieldMapping.id) {
     mappingModel.uuid = fieldMapping.id;
  }

  mappingModel.fieldMappings = [];
```

Keeping object instance naming consistent with its associated type allows for easer and quicker refactoring in the future, and a more intuitive debugging when necessary.

igarashitm commented:
@deeleman thank you for the review! so I believe two follow up commits addressed all of them. Can you approve it if it looks good?

deeleman commented on src/app/lib/atlasmap-data-mapper/services/mapping.management.service.ts:
```diff
@@ -91,35 +88,29 @@ export class MappingManagementService {
+                const operation = this.http.get(url).map((res:Response) => res.json());
                 operations.push(operation);
             }
             Observable.forkJoin(operations).subscribe((data:any[]) => {
```
Well, I'm not that familiar with this service and its overall purpose, so for me it's a bit hard to understand all its nuances, but please bear in mind that this subscription will remain open even when the service is disposed by the Angular dep injection system (which is usually at the end of the application lifespan, so it's not a big deal after all). If that is okay for you, then go ahead.

Otherwise, you can always desubscribe from it after certain condition applies. Usually it is not a good idea to leave RxJS subscriptions open in the wild, but this is a service and services are meant to persist, including its logic, so I leave this call up to you ;)

igarashitm replied:
Filed an issue for this one, I think we need to check if there're same issues elsewhere in the code and fix them:
https://github.com/atlasmap/atlasmap-ui/issues/130

igarashitm commented:
cleaned up and changed the condition a bit here,
https://github.com/atlasmap/atlasmap-ui/pull/129/files#diff-7d660b499eb52f76d0c8d7b20bade29cR316

So if scope or id is not contained in the validation result, just show them in the global error area anyway. It can't be traced to specific mapping though, still some outcome to show.
